### ISSUE TITLE
Cov not updating coverage on MacOS

### ISSUE THREAD

**Issue by thoni56** [AdamNiederer/cov#43]
I'm running Emacs for MacOS 27.2 and it seems that Emacs does not notice the updated `.gcov` files.

If I touch a source file Emacs says "Reverting ..." as it should. But if I modify the `.gcov` file, this does not trigger an update of the coverage in the fringe. I need to re-visit the file, then it is updated. I don't know if they even use the same mechanism for watching for file changes, so maybe that is a red herring. Or a clue...

**Comment by thoni56** [comment]
Ok, maybe [this ](https://emacs.stackexchange.com/questions/3866/file-notify-support-on-osx) is the reason.

**Comment by snogge** [comment]
Could be.  As far as I can tell from both the stack overflow question and the elisp manual, `kqueue` should be supported by  filenotify.el.  But I do not have access to any MacOS machines to test on. 
Maybe you can check whether your emacs has the kqueue support enabled?
I think autorevert uses file polling if no notification support is available, so it might not be an indication of kqueue support.

**Comment by snogge** [comment]
The only way for cov.el to discover changes in the gcov file is file-notify.
cov checks whether `file-notify` is supported by checking the value of `file-notify--library`, you can check that variable in your emacs to see whether `file-notify` was enabled in your build.
`cov-update` will only update the overlays from coverage data already stored in `cov-coverages`, loading coverage data only if it is not already present.
If you first do 
```
(setq cov-coverages (make-hash-table :test 'equal))
```
and then `cov-udpate`, the coverage data should be updated. Of course this will delete loaded coveage data for all files, but any existing overlays will remain in place.
This is probably an unintended consequence (bug) from when the `file-notify` support was added to cov.el.

**Comment by thoni56** [comment]
`file-notify--library` is `kqueue` both for GUI Emacs and Homebrew Emacs.

I've updated everything to the latest version, cov-mode (mine was one release old), Emacs and Monterey. Still no automatic cov-updating, but at least a manual `cov-update` consistently refreshes the fringe as it is supposed to. So that is some small progress.

I tried to see if `file-notify` works as advertised on MacOS using

    (file-notify-add-watch "/tmp/x" '(change change) 'my-notify-callback)

and then update the content of `/tmp/x`. That seemed to result in the correct events.

But what I discovered was that if the file is removed a "stopped" event is generated. And of course then no more events were sent. I don't know if this is different from other environments, but if it is, might that be a problem?

**Comment by thoni56** [comment]
Nope. `inotify` has the same behaviour (sending "stopped" event when file is removed)...

**Comment by thoni56** [comment]
Aaahh! There is no problem with updating the actual coverage. It is the fringe that is not updated on change in coverage!

Not until I make an edit... If I have a "new" coverage, which has not been shown in the fringe yet, all I have to do is to make a single character press, and the fringe is updated to show the new coverage.

That makes it a bit easier to live with, but still a nuisance.

Perhaps this also makes it easier to figure out what's different with MacOS Emacsen.

**Comment by snogge** [comment]
Do I understand correctly that this (the delayed update of the fringe) is only a problem on MacOS?

**Comment by thoni56** [comment]
Yes, as far as I can see. I'll try to get in touch with maintainers of Emacs for MacOS to see if they have any clues about the fringe problem.

**Comment by thoni56** [comment]
Looks like this is not related to `cov-mode`. It is probably caused by some spurious error that seems to originate from `org-mode` : `Selecting deleted buffer`. I think `org-mode` is probably also hooked into `file-notify` and the error message prevents `cov-mode` to do its thing.

This is mostly guesswork from my part based on some googling the error message (indicating that one possible source could be `org-mode`) and then noticing that if I did not open any `org-mode` file I couldn't see this happening. Also an experiment on WSL confirmed that I could trigger this error if I first opened my TODO-org-mode file.

So closing. Sorry for the noise.
